enterprise/backend/src/metabase_enterprise/task/caching.clj

```diff
@@ -0,0 +1,92 @@
+  "Calculate when a next run should happen based on a cron schedule"
+  [^String schedule]
+  (let [^MutableTrigger cron (cron/finalize (cron/cron-schedule schedule))]
+    ;; Synchronize cron runner to our app time, which may be mocked in tests
+    (.setStartTime cron (t/java-date))
+    (-> (.getFireTimeAfter cron (t/java-date (t/offset-date-time)))
+        (t/offset-date-time (t/zone-offset)))))
+
+(defn- refresh-schedule-configs
+  "Update `last_expired_at` for every cache config with `:schedule` strategy"
+  []
+  (count
+   (for [{:keys [id config]} (select-ready-to-run :schedule)]
+     (t2/update! :model/CacheConfig {:id id}
+                 {:next_run_at     (calc-next-run (:schedule config))
+                  :last_expired_at (t/offset-date-time)}))))
+
+(defn- refresh-query-configs
+  "Fetches `:query`-strategy configs wants to re-check their queries, runs those queries and updates `last_expired_at`
+  where `(:marker state)` is not equal to the result of the query."
+  []
+  (if-let [configs (seq (select-ready-to-run :query))]
+    (let [fields (m/index-by :id (t2/select :model/Field :id [:in (map #(-> % :config :field_id) configs)]))
+          tables (m/index-by :id (t2/select :model/Table :id [:in (map :table_id (vals fields))]))]
+      (count
+       (for [{:keys [id config] :as item} configs
+             :let [field       (get fields (:field_id config))
+                   table       (get tables (:table_id field))
+                   query       {:database (:db_id table)
+                                :type     :query
+                                :query    {:source-table (:table_id field)
+                                           :aggregation  [(:aggregation config) [:field (:id field) nil]]}}
+                   result      (-> (qp/process-query query) :data :rows ffirst)
+                   next-run-at (calc-next-run (:schedule config))
+                   marker      (:marker (:state item))]]
+         (t2/update! :model/CacheConfig {:id id}
+                     (cond-> {:next_run_at next-run-at}
+                       (not= marker result) (assoc :state {:marker result}
+                                                   :last_expired_at (t/offset-date-time)))))))
+    0))
+
+(jobs/defjob ^{org.quartz.DisallowConcurrentExecution true
+               :doc                                   "Refresh 'schedule' caches"}
+  Caching [_ctx]
+  (refresh-schedule-configs)
+  (refresh-query-configs))
+
+(def ^:private caching-job
+  (jobs/build
+    (jobs/with-description "Schedule Caches refresh task")
+    (jobs/of-type Caching)
+    (jobs/with-identity (jobs/key "metabase-enterprise.Caching.job"))
+    (jobs/store-durably)))
+
+(def ^:private caching-trigger
+  (triggers/build
+    (triggers/with-identity (triggers/key "metabase-enterprise.Caching.trigger"))
+    (triggers/start-now)
+    (triggers/with-schedule
+      (cron/schedule
+        ;; run every minute
+        (cron/cron-schedule "0 * * * * ? *")))))
```

Comment:
Is this too frequent?
- for scheduled cache: what's the lowest unit users can configure on UI? (aside: for pulse people can only configure to the hours, there is a discussion to make it more granular [here](https://github.com/metabase/metabase/issues/3846)
- For query cache: I can see how we want to run it more often for query cache, but 1m might be too frequent IMO. `max` on a indexed column is optimized, but `count` is a full table scan.

Reply:
okay, I just realized this is just to trigger the check, but my concern for performance still stands; we're giving users the ability to configure using cronjob syntax, and it's possible that they will make something run every minute. This comes with a significant performance impact.

What's your thought on this?

cc @luizarakaki

Reply:
How does the scheduler work, is there backpressure if the jobs start taking more than a minute?

Reply:
Well it's not going to run every query config every minute, it just selects whatever is ready to be refreshed once a minute. I mean you can configure stuff in a weird way, but once a minute also gives you nice distribution - you won't run all your queries at the same time all the time, IMO.

I'm pretty sure it comes with negligible performance impact since you execute 2 queries (1 for schedule/1 for query strategies) once a minute. And usually they return nothing.

About backpressure: not sure, but I'm pretty sure if we have a problem with backpressure/performance/dropping tasks, it should be done separately.

Reply:
Re: backpressure, I was just curious and this felt like a good time to ask. Like this pattern as a whole.

Reply:
We actually seem to have [some trouble about execution](https://metaboat.slack.com/archives/C05MPF0TM3L/p1704727823706109) when it spills over the time difference - I've digged a bit, but couldn't get through.

But it's really separate IMO.